include/rfb.js

```diff
@@ -507,23 +502,25 @@ var RFB;
                     } else if (oldstate === 'init') {
                         Util.Error("Error while initializing.");
                     }
-
-                    // Make sure we transition to disconnected
-                    setTimeout(function () {
-                        this._updateState('disconnected');
-                    }.bind(this), 50);
-
                     break;
 
                 default:
                     // No state change action to take
             }
 
             if (oldstate === 'failed' && state === 'disconnected') {
+                // do disconnect action, but stay in failed state and
+                // keep the previous status message
+                this._rfb_state = 'failed';
                 this._onUpdateState(this, state, oldstate);
             } else {
                 this._onUpdateState(this, state, oldstate, statusMsg);
             }
+
+            // Make sure we transition to disconnected
+            if (state === 'failed') {
+                this._updateState('disconnected');
+            }
```

Comment:
shouldn't we wait until actually receiving the "disconnected" event from the WebSocket (it's late, so I may be missing something here)?